### ISSUE TITLE
design discussion - handling arbitrary components as arbitrary properties

### ISSUE THREAD

**Issue by chriddyp** [plotly/dash-renderer#95]

**Comment by chriddyp** [comment]
This issue is to discuss how to rendering arbitrary components as arbitrary properties in Dash. This has been attempted a few times in the past, e.g. #32 and #92 .

**Current State of the World**

In the current version of Dash, `children` is the only property that can contain a Dash component. The `dash` layout is a tree that can only branch through `children`.

This constraint on the data structure is used in the following places (perhaps more):
- `TreeContainer` recurses through `children`
- The base component in `dash` (`dash.development.base_component.Component`) recurses through `children`: https://github.com/plotly/dash/blob/1663feaf7529536d29def59f29f14f8ac2ac43fa/dash/development/base_component.py#L180-L191. This layout crawling is used in a few places for validation (e.g. checking if an ID exists in the tree).

**Ideal API**

As mentioned in #92, if we support rendering components as properties, then these components should be first-class citizens of dash, just like the children component. That is:
1. The property can be updated via an `Output` callback 
2. The property can be an `Input` or a `State`
3. The component could contain components of its own (as `children` or as other properties) and those nested components are all first-class
4. The property can be a single component, a list of components, or a list of mixed types
5. Component could be deeply nested within a list or a dict. For example:
```
DataTable(
   cell_renderer=[
        {'columnA': dcc.Dropdown(...), 'columnB': dcc.Dropdown(...), ...},
        ...
   ]
)
```

**Challenges**

Item 5 above would make things difficult with our current architecture. Frequently, the `layout` can become pretty huge: consider a `dcc.Graph` with 1M points. With our current architecture (recursively crawling the layout), to support nested components, we'd need to:
- Crawl the entire 1M points, as we wouldn't know apriori whether or not an item within the array might contain a component.
- Figure out some way to skip crawling certain keys, perhaps via the `metadata.json` file. In practice, `dcc.Graph.figure` can't contain any components (as plotly.js doesn't render arbitrary React components), so we don't need to waste the effort in crawling it. However, other components might have mixed values: they could contain either values _or_ components (e.g. `dash_table.DataTable.data`). In those cases, there would be no getting around needing to crawl the entire figure.
- Change our architecture so that we don't have to crawl the tree.

Now, I might be pre-optimizing things here. Maybe crawling every key in the layout isn't actually that big of a deal. After we crawl, we keep cache the location of component IDs within the layout in the `paths` store, so we might only need to do this once.

However, these types of requirements make me wonder if our recursive renderer is actually the right foundation. An alternative architecture that I've thought about would be something like this:


```javascript
/* A sample Dash component code */
class Item extends Component {
    shouldComponentUpdate(nextProps) {
        // Since our tree data is immutable (via ramda)
        // we can do easy inequality checks like this to
        // determine if the sub-tree has changed or not.
        // The nested `R.set(R.lensPath(...), newVal, existingTree)`
        // will update the references in the direct path from the top node
        // down to the node that got updated.
        // see this ramda example: https://goo.gl/4zQjMp
        return nextProps !== this.props;
    }

    render() {
        // props provided by dash-renderer
        const {connect, setProps} = this.props;

        // user-provided properties
        const {children, header} = this.props;
        return (
            <div>
                {connect(header)}
                <input onChange={e => setProps({value: e.target.value})}

                {connect(children)}
            </div>
        )
    }
}
```

And this is roughly what `dash-renderer` could look like. 
No redux here, just sketching things out with `this.state`.
```javascript
function TreeContainer(props) {
    constructor(props) {
        this.state = {layout: props.layout}
        this.paths = computePathsByCrawlingLayout(props.layout)
    }

    createSetProps(id) {
        return newProps => {
            const pathOfComponentProps = R.lensPath(R.append(
                'props',
                this.state.paths[id]
            ));
            const existingProps = R.path(pathOfComponentProps, this.state.layout);
            const mergedProps = R.merge(existingProps, newProps);
            this.setState({
                layout: R.set(pathOfComponentProps, mergedProps)
            });

            // fire off async updates
            this.notifyObservers(id, newProps);
        }
    }

    connect(maybeComponent) {
        if (R.type(maybeComponent) === 'Object' &&
            R.has('type', maybeComponent) &&
            R.has('namespace', maybeComponent)
        ) {
            const boundSetProps = this.createSetProps(maybeComponent.props.id);

            // connect is a "render"
            return React.createElement(
                window[maybeComponent.namespace][maybeComponent.type],
                ...R.merge(
                    maybeComponent.props,
                    {
                        setProps: boundSetProps,
                        connect: this.connect
                    }
                )
            );
        }
    }

    notifyObservers(id, newProps) {
        // ... create the payloads, etc etc
        fetch(`/_dash-update-component`, payload).then(res => {
            // crawl the entire response to see if there are any components 
            // with IDs. if so, update the paths mapping 
            // (the mapping between component ID and component location in the tree)
            // I don't think that there is any way to get around needed to crawl this tree
            updatedPaths = computePathsByCrawlingLayout(subTree)

            // paths doesn't need to be in the state since it's not rendered
            // keeping it out of the state prevents unecessary updates on this.setState
            this.paths = updatedPaths;
            
            this.setState({layout: updatedLayout});
        })
    }

    render() {
        return connect(props);
    }

}
```

With that sketch, its up to the component author to use `connect` appropriately. If they want to expose a Dash component as a property (no matter how nested), then they can just wrap that property in `connect`. No more recursion - we're moving the rendering into the `connect`.

> There's a few other nice things that I included in that sketch above:
> - No redux
> - A path towards efficient rendering via a simple shouldComponentUpdate. Or, equivalently, all components could inherit from `PureComponent`

Of course, this method introduces a new parameter (`connect`) and so it'd be a breaking change for component authors.

I'm not sure if we can do this in a way that doesn't break things for component authors. Perhaps there would be a way to keep `children` "automatically connected" and then the `connect` prop would just be something additive to the API. I haven't thought this through yet.

***

In addition to reworking dash-renderer, we'd also need to rework `dash.development.base_component.Component` or deprecate it. I doubt that there are many folks besides us that are using methods like `.traverse` or `.keys()`.

***

Another way forward would be to ditch requirement 5 and only handle components as shallow properties: no nested components. I believe that this would be do-able with our current architecture (no breaking changes) and we'd still get a _ton_ of value out of it. It would essentially be #92 but with the additional `actions` stuff (taking into account `paths` and the update pathways via callbacks). We'd still need to patch `dash.development.base_component.Component` to search for component IDs in properties other than `children`, but that shouldn't be too bad either.

**Comment by T4rk1n** [comment]
I must admit I didn't think of the usage of the components props as Input or Output in #92. They are wrapped in NotifyObservers like the rest of the layout so I don't think it needs a connect props, maybe just need to be supported elsewhere in the renderer chain of actions and also supported in dash callback validation.

**Comment by chriddyp** [comment]
>  so I don't think it needs a connect props

Right, I think we'd only need `connect` if we want to support item 5 above - components that are nested in arbitrary data structures. We can forgo that requirement right now and fit it into the existing architecture.
